std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,111 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range.
+This function has $(D O(r.length)) complexity.
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+auto mean(R)(R r) if (isInputRange!R && !isInfinite!R)
+in
+{
+    static if (hasLength!R)
+    {
+        assert(r.length > 0);
+    }
+}
+body
+{
+    import std.range.primitives : hasLength;
+
+    static if (hasLength!R)
+    {
+        return sum(r) / r.length;
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        immutable auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))(tuple(0u, 0.0), r);
+        return pair[1] / pair[0];
+    }
+}
+
+///
+@safe unittest {
+    import std.math : approxEqual;
+
+    assert(mean([1, 2, 3]) == 2);
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
+}
+
+unittest {
+    import std.internal.test.dummyrange : ReferenceInputRange;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(approxEqual(r2.mean, 5.5));
+}
+
+/**
+Finds the median of a random access range. This function assumes
+r is an already sorted range. See $(D assumeSorted) for details.
+This function is $(D Ο(1)) in release mode, and $(D O(log(r.length)))
+in debug mode.
+
+Params:
+    r = a random access range
+
+Returns:
+    the median of r
+*/
+auto median(R)(R r)
+if (isRandomAccessRange!R &&
+    !isInfinite!R)
+in
+{
+    assert(r.length > 0);
+}
+body
+{
+    import std.range : assumeSorted;
+
+    if (r.length == 1)
+    {
+        return r[0];
+    }
+    else if (r.length == 2)
+    {
+        return (r[0] + r[1]) / 2;
+    }
+
+    auto sorted = assumeSorted(r);
```

Comment:
Uhm, this is wrong. `assumeSorted` should be in *user code*, not here. This function should take `SortedRange` as parameter, so that the user has to either sort his range, or call `assumeSorted`, before this function will accept the range. Putting `assumeSorted` here completely defeats the purpose of `SortedRange`.